beacon-chain/verification/cache.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		return err
+	}
+	sr, err := signing.ComputeSigningRootForRoot(sig.Root, domain)
+	if err != nil {
+		return err
+	}
+	if !s.Verify(pb, sr[:]) {
+		return signing.ErrSigFailedToVerify
+	}
+
+	return nil
+}
+
+// SignatureVerified checks the signature cache for the given key, and returns a boolean value of true
+// if it has been seen before, and an error value indicating whether the signature verification suceeded.
+// ie only a result of (true, nil) means a previous signature check passed.
+func (c *sigCache) SignatureVerified(sig SignatureData) (bool, error) {
+	val, seen := c.Get(sig)
+	if !seen {
+		return false, nil
+	}
+	verified, ok := val.(bool)
+	if !ok {
+		log.WithFields(sig.logFields()).Debug("ignoring invalid value found in signature cache")
+		// This shouldn't happen, and if it does, the caller should treat it as a cache miss and run verification
+		// again to correctly populate the cache key.
+		return false, nil
+	}
+	if verified {
+		return true, nil
+	}
+	return true, signing.ErrSigFailedToVerify
+}
+
+// ProposerCache represents a type that can compute the proposer for a given slot + parent root,
+// and cache the result so that it can be reused when the same verification needs to be performed
+// across multiple values.
+type ProposerCache interface {
+	ComputeProposer(ctx context.Context, root [32]byte, slot primitives.Slot, pst state.BeaconState) (primitives.ValidatorIndex, error)
+	Proposer(root [32]byte, slot primitives.Slot) (primitives.ValidatorIndex, bool)
+}
+
+// ProposerData represents the set of inputs, aside from the parent beacon state,
+// that identify a query for a proposer index. This is used as the proposer index cache key.
+type ProposerData struct {
+	Parent [32]byte
+	Slot   primitives.Slot
+}
+
+func (d ProposerData) logFields() logrus.Fields {
+	return logrus.Fields{
+		"parent_root": fmt.Sprintf("%#x", d.Parent),
+		"slot":        d.Slot,
+	}
+}
+
+func newPropCache(size int) *propCache {
+	return &propCache{Cache: lruwrpr.New(size)}
+}
+
+type propCache struct {
+	*lru.Cache
+}
+
+// ComputeProposer takes the state for the given parent root and slot and computes the proposer index, updating the
+// proposer index cache when successful.
+func (c *propCache) ComputeProposer(ctx context.Context, parent [32]byte, slot primitives.Slot, pst state.BeaconState) (primitives.ValidatorIndex, error) {
+	pst, err := transition.ProcessSlotsUsingNextSlotCache(ctx, pst, parent[:], slot)
+	if err != nil {
+		return 0, err
+	}
+	idx, err := helpers.BeaconProposerIndex(ctx, pst)
+	if err != nil {
+		return 0, err
+	}
+	c.Add(ProposerData{Parent: parent, Slot: slot}, idx)
+	return idx, nil
+}
+
+// Proposer returns the validator index if it is found in the cache, along with a boolean indicating
+// whether the value was present, similar to accessing an lru or go map.
+func (c *propCache) Proposer(parent [32]byte, slot primitives.Slot) (primitives.ValidatorIndex, bool) {
```

Comment:
we probably want to support this using a different cache (potuz' bundle cache)

Reply:
It is a weird API to take the parent root here... But also we already have a proposer ID cache, I don't understand why we are using a fully new one here.